sql/group.go

```diff
@@ -0,0 +1,438 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package sql
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/sql/parser"
+	"github.com/cockroachdb/cockroach/util/log"
+)
+
+var aggregates = map[string]aggregateFunc{
+	"avg":   &avgAggregate{},
+	"count": &countAggregate{},
+	"max":   &maxAggregate{},
+	"min":   &minAggregate{},
+	"sum":   &sumAggregate{},
+}
+
+func (p *planner) groupBy(n *parser.Select, s *scanNode) (*groupNode, error) {
+	// TODO(pmattis): This only handles aggregate functions, not GROUP BY.
+
+	// Loop over the render expressions and extract any aggregate functions.
+	var funcs []*parser.FuncExpr
+	for _, r := range s.render {
+		f, err := extractAggregateFuncs(r)
+		if err != nil {
+			return nil, err
+		}
+		funcs = append(funcs, f...)
+	}
+	if len(funcs) == 0 {
+		return nil, nil
+	}
+
+	// Aggregation is being performed. Loop over the render expressions again and
+	// verify that the only qvalues mentioned by GROUP BY expressions are allowed
+	// outside of the aggregate function arguments. For example, the following is
+	// illegal because k is used outside of the aggregate function and not part
+	// of a GROUP BY expression.
+	//
+	//   SELECT COUNT(k), k FROM kv GROUP BY v
+	for _, r := range s.render {
+		if err := checkAggregateExpr(r); err != nil {
+			return nil, err
+		}
+	}
+
+	if log.V(2) {
+		log.Infof("Group: %s", funcs)
+	}
+
+	group := &groupNode{
+		columns: s.columns,
+		render:  s.render,
+		funcs:   funcs,
+		aggs:    make([]aggregateFunc, 0, len(funcs)),
+	}
+
+	for _, f := range funcs {
+		agg, ok := findAggregateFunc(f)
+		if !ok {
+			panic(fmt.Sprintf("%s not found (this shouldn't happen)", f.Name))
+		}
+		if agg == nil {
+			return nil, fmt.Errorf("%s not implemented", f.Name)
+		}
+		group.aggs = append(group.aggs, agg.New())
+	}
+
+	// Replace the render expressions in the scanNode with expressions that
+	// compute only the arguments to the aggregate expressions.
+	s.columns = make([]string, 0, len(funcs))
+	s.render = make([]parser.Expr, 0, len(funcs))
+	for _, f := range funcs {
+		if len(f.Exprs) != 1 {
+			panic(fmt.Sprintf("%s has %d arguments (expected 1)", f.Name, len(f.Exprs)))
+		}
+		s.columns = append(s.columns, f.String())
```

Comment:
The columns get passed back in the response and up to the driver. They appear in the `Rows.Columns` result.